synthetic_signwriting/synthetic_signwriting.py

```diff
@@ -1,37 +1,86 @@
         self.faces = []
 
-    def add_keyframe(self, face=True, left_hand=True, right_hand=True):
+    def add_keyframe(self, face=None, left_hand=None, right_hand=None):
         # Create and configure a new frame
-        frame = self.pose.create_frame()
-        if face:
-            frame.set_face(...)
-        if left_hand:
-            frame.set_left_hand(...)
-        if right_hand:
-            frame.set_right_hand(...)
+        self.hands.append((left_hand, right_hand))
+        self.faces.append(face)
+
+    def add_random_keyframe(self):
+        # choose a random hand and face pose
+        right_hand = randomize_hand_pose(self.hand_data)  # choose a random hand pose
+        left_hand = randomize_hand_pose(self.hand_data)  # choose a random hand pose
+
+        # choose a random face pose
+        face_type = random.choice(list(self.face_data.keys()))
+        element = random.choice(list(self.face_data[face_type].keys()))
+        if element == "source":
+            face = self.face_data[face_type][element]
+        else:
+            cathartic = random.choice(list(self.face_data[face_type][element].keys()))
+            face = self.face_data[face_type][element][cathartic]
+
+        # add the keyframe
+        self.add_keyframe(face=face, left_hand=left_hand, right_hand=right_hand)
 
     def render_signwriting(self):
         # Use the list of hands and faces to generate SignWriting
         return "..."
 
-    def render_pose(self):
-        self._smooth()
-        # Apply camera transformation and render the pose sequence
-        return "..."
+    def render_pose(self, frame_suspension=100):
+        self.pose.body.data = None
+        zero_frame = np.zeros((frame_suspension - 2, 1, 543, 3))  # the -2 is to remove the first and last frame
+
+        for face, (left_hand, right_hand) in zip(self.faces, self.hands):
+            selected_frame = random.choice(self.static_position)  # choose a random base frame
+            pose_frame = self.static_position[selected_frame][0].copy()
+
+            # hand
+            right_hand_base = pose_frame[522:543]  # get the right hand base that located (522:543)
+            right_hand = prorate_hand(right_hand, right_hand_base)
+            pose_frame[522:543] = right_hand
+            left_hand_base = pose_frame[501:522]  # get the left hand base that located (501:522)
+            left_hand = prorate_hand(left_hand, left_hand_base, reflection=True)  # reflection because it's left hand
+            pose_frame[501:522] = left_hand
+
+            # face
+            face_base = pose_frame[33:501]  # get the face base that located (33:501)
+            face = prorate_face(face, face_base)
+            pose_frame[33:501] = face
 
-    def _smooth(self, kind=None):
-        if kind is None:
-            kind = random.choice(["linear", "quadratic", "cubic"])
+            if self.pose.body.data is None:
+                self.pose.body.data = [[create_base_frame(pose_frame)]]
```

Comment:
i see why you have issues...

what you are trying to do is:

```py
frames = [] # in the beginning
frames.append(....) # inside the loop
# after
pose.body = NumPyPoseBody(fps=pose.body.fps, data=np.array(frames))
```

the thing is that the pose body also has "confidence" information. you could start with this, and we could add the confidence later (0 or 1)

then `interpolate` would also work just fine

Reply:
so to change for now your smooth technique keep it the same I didn't understand?